File: lib/utils/fluxo-sequencial-turno.ts
import { criarPosicaoFilaSetor } from './capacidade-setor.ts'
import type {
  SnapshotParcelamentoDemandaTurnoV2,
  TurnoSetorDemandaStatusV2,
  TurnoSetorFilaStatusV2,
} from '@/types'

export interface DemandaFluxoSequencialBase {
  id: string
  turnoOpId: string
  setorId: string
  setorCodigo: number
  setorNome: string
  quantidadePlanejada: number
  quantidadeRealizada: number
  status: TurnoSetorDemandaStatusV2
  iniciadoEm: string | null
  encerradoEm: string | null
}

export interface DiagnosticoFluxoSequencialDemandaV2 {
  posicaoFila: number | null
  statusFila: TurnoSetorFilaStatusV2
  quantidadeBacklogSetor: number
  quantidadeAceitaTurno: number
  quantidadeExcedenteTurno: number
  quantidadePendenteSetor: number
  quantidadeLiberadaSetor: number
  quantidadeDisponivelApontamento: number
  quantidadeBloqueadaAnterior: number
  setorAnteriorId: string | null
  setorAnteriorCodigo: number | null
  setorAnteriorNome: string | null
}

function normalizarInteiroNaoNegativo(valor: number): number {
  if (!Number.isFinite(valor) || valor <= 0) {
    return 0
  }

  return Math.floor(valor)
}

function compararDemandasPorFluxo(
  primeiraDemanda: DemandaFluxoSequencialBase,
  segundaDemanda: DemandaFluxoSequencialBase
): number {
  if (primeiraDemanda.setorCodigo !== segundaDemanda.setorCodigo) {
    return primeiraDemanda.setorCodigo - segundaDemanda.setorCodigo
  }

  return primeiraDemanda.setorId.localeCompare(segundaDemanda.setorId)
}

export function calcularQuantidadeLiberadaSetor(
  quantidadePlanejada: number,
  quantidadeConcluidaSetorAnterior?: number | null
): number {
  const quantidadePlanejadaNormalizada = normalizarInteiroNaoNegativo(quantidadePlanejada)

  if (quantidadeConcluidaSetorAnterior === null || quantidadeConcluidaSetorAnterior === undefined) {
    return quantidadePlanejadaNormalizada
  }

  return Math.min(
    quantidadePlanejadaNormalizada,
    normalizarInteiroNaoNegativo(quantidadeConcluidaSetorAnterior)
  )
}

export function calcularQuantidadeDisponivelApontamento(
  quantidadePlanejada: number,
  quantidadeRealizadaAtual: number,
  quantidadeConcluidaSetorAnterior?: number | null
): number {
  const quantidadeLiberadaSetor = calcularQuantidadeLiberadaSetor(
    quantidadePlanejada,
    quantidadeConcluidaSetorAnterior
  )

  return Math.max(
    quantidadeLiberadaSetor - normalizarInteiroNaoNegativo(quantidadeRealizadaAtual),
    0
  )
}

export function criarSnapshotParcelamentoDemandaTurno(input: {
  quantidadePlanejada: number
  quantidadeRealizadaAtual: number
  quantidadeDisponivelApontamento: number
}): SnapshotParcelamentoDemandaTurnoV2 {
  const quantidadeBacklogSetor = Math.max(
    normalizarInteiroNaoNegativo(input.quantidadePlanejada) -
      normalizarInteiroNaoNegativo(input.quantidadeRealizadaAtual),
    0
  )
  const quantidadeAceitaTurno = Math.min(
    quantidadeBacklogSetor,
    normalizarInteiroNaoNegativo(input.quantidadeDisponivelApontamento)
  )

  return {
    quantidadeBacklogSetor,
    quantidadeAceitaTurno,
    quantidadeExcedenteTurno: Math.max(quantidadeBacklogSetor - quantidadeAceitaTurno, 0),
  }
}

export function enriquecerDemandasComFluxoSequencial<T extends DemandaFluxoSequencialBase>(
  demandas: T[]
): Array<T & DiagnosticoFluxoSequencialDemandaV2> {
  const demandasPorTurnoOp = new Map<string, T[]>()

  for (const demanda of demandas) {
    const demandasTurnoOp = demandasPorTurnoOp.get(demanda.turnoOpId) ?? []
    demandasTurnoOp.push(demanda)
    demandasPorTurnoOp.set(demanda.turnoOpId, demandasTurnoOp)
  }

  const diagnosticosPorDemandaId = new Map<string, Omit<DiagnosticoFluxoSequencialDemandaV2, 'posicaoFila' | 'statusFila'>>()

  for (const demandasTurnoOp of demandasPorTurnoOp.values()) {
    const demandasOrdenadas = [...demandasTurnoOp].sort(compararDemandasPorFluxo)

    demandasOrdenadas.forEach((demandaAtual, indiceAtual) => {
      const demandaAnterior = indiceAtual > 0 ? demandasOrdenadas[indiceAtual - 1] : null
      const quantidadePendenteSetor = Math.max(
        normalizarInteiroNaoNegativo(demandaAtual.quantidadePlanejada) -
          normalizarInteiroNaoNegativo(demandaAtual.quantidadeRealizada),
        0
      )
      const quantidadeLiberadaSetor = calcularQuantidadeLiberadaSetor(
        demandaAtual.quantidadePlanejada,
        demandaAnterior?.quantidadeRealizada ?? null
      )
      const quantidadeDisponivelApontamento = calcularQuantidadeDisponivelApontamento(
        demandaAtual.quantidadePlanejada,
        demandaAtual.quantidadeRealizada,
        demandaAnterior?.quantidadeRealizada ?? null
      )

      const snapshotParcelamento = criarSnapshotParcelamentoDemandaTurno({
        quantidadePlanejada: demandaAtual.quantidadePlanejada,
        quantidadeRealizadaAtual: demandaAtual.quantidadeRealizada,
        quantidadeDisponivelApontamento,
      })

      diagnosticosPorDemandaId.set(demandaAtual.id, {
        quantidadeBacklogSetor: snapshotParcelamento.quantidadeBacklogSetor,
        quantidadeAceitaTurno: snapshotParcelamento.quantidadeAceitaTurno,
        quantidadeExcedenteTurno: snapshotParcelamento.quantidadeExcedenteTurno,
        quantidadePendenteSetor,
        quantidadeLiberadaSetor,
        quantidadeDisponivelApontamento,
        quantidadeBloqueadaAnterior: Math.max(
          quantidadePendenteSetor - quantidadeDisponivelApontamento,
          0
        ),
        setorAnteriorId: demandaAnterior?.setorId ?? null,
        setorAnteriorCodigo: demandaAnterior?.setorCodigo ?? null,
        setorAnteriorNome: demandaAnterior?.setorNome ?? null,
      })
    })
  }

  const posicoesPorDemandaId = new Map<string, number | null>()
  const demandasPorSetor = new Map<string, T[]>()

  for (const demanda of demandas) {
    const demandasSetor = demandasPorSetor.get(demanda.setorId) ?? []
    demandasSetor.push(demanda)
    demandasPorSetor.set(demanda.setorId, demandasSetor)
  }

  for (const demandasSetor of demandasPorSetor.values()) {
    const filaAtiva = demandasSetor.filter(
      (demanda) => demanda.status !== 'concluida' && demanda.status !== 'encerrada_manualmente'
    )

    filaAtiva.forEach((demanda, indice) => {
      posicoesPorDemandaId.set(demanda.id, indice + 1)
    })
  }

  return demandas.map((demanda) => {
    const diagnostico = diagnosticosPorDemandaId.get(demanda.id)
    const quantidadePendenteSetor = diagnostico?.quantidadePendenteSetor ?? 0
    const quantidadeDisponivelApontamento = diagnostico?.quantidadeDisponivelApontamento ?? 0
    const snapshotParcelamento = criarSnapshotParcelamentoDemandaTurno({
      quantidadePlanejada: demanda.quantidadePlanejada,
      quantidadeRealizadaAtual: demanda.quantidadeRealizada,
      quantidadeDisponivelApontamento,
    })
    const posicaoFila = demanda.encerradoEm ? null : (posicoesPorDemandaId.get(demanda.id) ?? null)
    const fila = criarPosicaoFilaSetor({
      quantidadePlanejada: demanda.quantidadePlanejada,
      quantidadeConcluida: demanda.quantidadeRealizada,
      quantidadeDisponivelApontamento,
      posicaoFila,
      iniciadoEm: demanda.iniciadoEm,
      encerradoEm: demanda.encerradoEm,
    })

    return {
      ...demanda,
      posicaoFila: fila.posicaoFila,
      statusFila: fila.statusFila,
      quantidadeBacklogSetor: snapshotParcelamento.quantidadeBacklogSetor,
      quantidadeAceitaTurno: snapshotParcelamento.quantidadeAceitaTurno,
      quantidadeExcedenteTurno: snapshotParcelamento.quantidadeExcedenteTurno,
      quantidadePendenteSetor,
      quantidadeLiberadaSetor: diagnostico?.quantidadeLiberadaSetor ?? 0,
      quantidadeDisponivelApontamento,
      quantidadeBloqueadaAnterior: diagnostico?.quantidadeBloqueadaAnterior ?? 0,
      setorAnteriorId: diagnostico?.setorAnteriorId ?? null,
      setorAnteriorCodigo: diagnostico?.setorAnteriorCodigo ?? null,
      setorAnteriorNome: diagnostico?.setorAnteriorNome ?? null,
    }
  })
}

File: backend/shared/fluxo_sequencial_turno.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Sequence

from shared.turno_dominio import normalizar_inteiro_nao_negativo


@dataclass(frozen=True)
class DemandaFluxoSequencialBase:
    id: str
    turno_op_id: str
    setor_id: str
    setor_sequencia_fluxo: int
    setor_nome: str
    quantidade_planejada: int
    quantidade_realizada: int
    quantidade_herdada_setor: int
    quantidade_liberada_setor: int
    status: str
    iniciado_em: object | None
    encerrado_em: object | None


@dataclass(frozen=True)
class DemandaFluxoSequencialEnriquecida(DemandaFluxoSequencialBase):
    quantidade_pendente_setor: int
    quantidade_liberada_setor_fluxo: int
    quantidade_disponivel_apontamento: int
    setor_anterior_id: str | None


def calcular_progresso_operacional_setor(
    demanda: DemandaFluxoSequencialBase,
) -> int:
    return min(
        normalizar_inteiro_nao_negativo(demanda.quantidade_planejada),
        normalizar_inteiro_nao_negativo(demanda.quantidade_realizada)
        + normalizar_inteiro_nao_negativo(demanda.quantidade_herdada_setor),
    )


def calcular_quantidade_liberada_setor(
    quantidade_planejada: int,
    quantidade_concluida_setor_anterior: int | None,
) -> int:
    quantidade_planejada_normalizada = normalizar_inteiro_nao_negativo(quantidade_planejada)
    if quantidade_concluida_setor_anterior is None:
        return quantidade_planejada_normalizada
    return min(
        quantidade_planejada_normalizada,
        normalizar_inteiro_nao_negativo(quantidade_concluida_setor_anterior),
    )


def enriquecer_demandas_com_fluxo_sequencial(
    demandas: Sequence[DemandaFluxoSequencialBase],
) -> list[DemandaFluxoSequencialEnriquecida]:
    demandas_por_turno_op: dict[str, list[DemandaFluxoSequencialBase]] = {}
    for demanda in demandas:
        demandas_por_turno_op.setdefault(demanda.turno_op_id, []).append(demanda)

    diagnosticos: dict[str, tuple[int, int, int, str | None]] = {}

    for demandas_turno_op in demandas_por_turno_op.values():
        demandas_ordenadas = sorted(
            demandas_turno_op,
            key=lambda demanda: (demanda.setor_sequencia_fluxo, demanda.setor_id),
        )
        for indice, demanda_atual in enumerate(demandas_ordenadas):
            demanda_anterior = demandas_ordenadas[indice - 1] if indice > 0 else None
            progresso_atual = calcular_progresso_operacional_setor(demanda_atual)
            progresso_anterior = (
                calcular_progresso_operacional_setor(demanda_anterior) if demanda_anterior else None
            )
            quantidade_pendente_setor = max(
                normalizar_inteiro_nao_negativo(demanda_atual.quantidade_planejada) - progresso_atual,
                0,
            )
            quantidade_liberada_setor = calcular_quantidade_liberada_setor(
                demanda_atual.quantidade_planejada,
                progresso_anterior,
            )
            quantidade_disponivel_apontamento = max(
                quantidade_liberada_setor - progresso_atual,
                0,
            )
            diagnosticos[demanda_atual.id] = (
                quantidade_pendente_setor,
                quantidade_liberada_setor,
                quantidade_disponivel_apontamento,
                demanda_anterior.setor_id if demanda_anterior else None,
            )

    enriquecidas: list[DemandaFluxoSequencialEnriquecida] = []
    for demanda in demandas:
        diagnostico = diagnosticos.get(demanda.id, (0, 0, 0, None))
        enriquecidas.append(
            DemandaFluxoSequencialEnriquecida(
                id=demanda.id,
                turno_op_id=demanda.turno_op_id,
                setor_id=demanda.setor_id,
                setor_sequencia_fluxo=demanda.setor_sequencia_fluxo,
                setor_nome=demanda.setor_nome,
                quantidade_planejada=demanda.quantidade_planejada,
                quantidade_realizada=demanda.quantidade_realizada,
                quantidade_herdada_setor=demanda.quantidade_herdada_setor,
                quantidade_liberada_setor=demanda.quantidade_liberada_setor,
                status=demanda.status,
                iniciado_em=demanda.iniciado_em,
                encerrado_em=demanda.encerrado_em,
                quantidade_pendente_setor=diagnostico[0],
                quantidade_liberada_setor_fluxo=diagnostico[1],
                quantidade_disponivel_apontamento=diagnostico[2],
                setor_anterior_id=diagnostico[3],
            )
        )
    return enriquecidas
